### DISCUSSION TITLE
Trying to send audio file via mdtest

### DISCUSSION THREAD

**Discussion by SoniaButtPunjabi** [tulir/whatsmeow#308] [WhatsApp protocol Q&A]
Hi Experts,

### I am newbie and a little know about coding and learning from searching internet, i need your help to guide me about my code
i want to send audio file via mdtest i edit main.go and insert this code

```
case "sendaud":
		if len(args) < 2 {
			log.Errorf("Usage: sendaud <jid> <audio path>")
			return
		}
		recipient, ok := parseJID(args[0])
		if !ok {
			return
		}
		data, err := os.ReadFile(args[1])
		if err != nil {
			log.Errorf("Failed to read %s: %v", args[0], err)
			return
		}
		uploaded, err := cli.Upload(context.Background(), data, whatsmeow.MediaAudio)
		if err != nil {
			log.Errorf("Failed to upload file: %v", err)
			return
		}
		msg := &waProto.Message{AudioMessage: &waProto.AudioMessage{
//			Seconds:       proto.Uint32(uint32(len(data))),	
//			StreamingSidecar: []byte{},
//			Waveform:      []byte{},
       			Url:           proto.String(uploaded.URL),
			Ptt:	       proto.Bool(true),
			DirectPath:    proto.String(uploaded.DirectPath),
			MediaKey:      uploaded.MediaKey,
       			Mimetype:      proto.String("audio/ogg; codecs=opus"),		                 	
       			FileEncSha256: uploaded.FileEncSHA256,
			FileSha256:    uploaded.FileSHA256,
			FileLength:    proto.Uint64(uint64(len(data))),		
		}}
		resp, err := cli.SendMessage(context.Background(), recipient, "", msg)
		if err != nil {
			log.Errorf("Error sending audio message: %v", err)
		} else {
			log.Infof("Image message sent (server timestamp: %s)", resp.Timestamp)
		}
	}
}
```

**This code successfully send voice message and play on Whatsapp web but not play on iphone and in android it's play but can't listen voice.**
#
Voice in .wav form

Please guide

**Comment by SoniaButtPunjabi** [comment]
@tulir please can you help me on this

**Comment by SoniaButtPunjabi** [comment]
Any expert here to guide?

**Comment by tulir** [comment]
Only .ogg with opus encoding is allowed for voice messages, .wav is not

**Comment by SoniaButtPunjabi** [comment]
Thanks for reply, is there any code for convert .wav to .ogg with opus
encoding in mdtest

**Reply by alipbudiman** [reply]
using ffmpeg

**Comment by niraj-tracewave** [comment]
@tulir 
below is our code for sending the audio in via S3 in the same format i.e ogg amd opus codecs, and the audio is sending perfectly fine but it not getting played.
Note: we've taken audioBytes by reading audio file which is uploaded on s3 bucket with the same file format [ogg]. Also audio is playable in web only not in device.

uploadResponse, uploadError := client.Upload(context.Background(), audioBytes, whatsmeow.MediaAudio)
		if uploadError != nil {
			log.Printf("Error uploading video: %s", uploadError.Error())
			return uploadError
		}	
		
msg = &waE2E.Message{
			AudioMessage: &waE2E.AudioMessage{
				Mimetype:          proto.String("audio/ogg; codecs=opus"),
				URL:               proto.String(uploadResponse.URL),
				DirectPath:        proto.String(uploadResponse.DirectPath),
				MediaKey:          uploadResponse.MediaKey,
				FileEncSHA256:     uploadResponse.FileEncSHA256,
				FileSHA256:        uploadResponse.FileSHA256,
				FileLength:        proto.Uint64(uploadResponse.FileLength),
				MediaKeyTimestamp: proto.Int64(time.Now().Unix()),
				PTT:               proto.Bool(true),
				Seconds:           proto.Uint32(duration),
				Waveform:          waveform,
			},
		}

**Reply by devlikepro** [reply]
here's ffmpeg worked for us, bit rate (or whatever the number if for) important too
```
ffmpeg -i input.mp3 -c:a libopus -b:a 32k -ar 48000 -ac 1 output.opus
```

**Reply by niraj-tracewave** [reply]
The above code is working perfectly for sending audio files. However, the issue is that while the audio plays correctly on the web, it does not play on mobile devices (iOS and Android). On devices, it shows an error saying, "Please ask the user to resend it."

**Reply by devlikepro** [reply]
yes, it accepts whatever data you send, try formatting your voice file with the command provided before sending it
